File: 2x2gamesolver.py
import tkinter as tk
import tkinter.messagebox
from functools import partial

# ----------------------- Setting up GUI --------------------------------
# Setting up home screen
window = tk.Tk()
window.title("2x2 simultaneous game solver")
canvas = tk.Canvas()
canvas.config(width = 450, height = 450)
canvas.create_rectangle(125, 125, 325, 325)
canvas.pack()
rowplayerlabel = tk.Label(text="Row\nplayer")
columnplayerlabel = tk.Label(text="Column\nplayer")
columnplayerlabel.place(x=210, y = 70)
rowplayerlabel.place(x=30, y = 200)
ulabel = tk.Label(text = "U")
ulabel.place(x=100, y = 165)
dplabel = tk.Label(text = "D")
dplabel.place(x=98, y = 265)
llabel = tk.Label(text = "L")
llabel.place(x=175, y = 95)
rlabel = tk.Label(text = "R")
rlabel.place(x=275, y = 95)
#Setting up payoff entry fields
ul1entry = tk.Entry(width=6)
ul1entry.place(x = 150, y = 190)
ul2entry = tk.Entry(width=6)
ul2entry.place(x = 180, y = 160)
ur1entry = tk.Entry(width=6)
ur1entry.place(x = 250, y = 190)
ur2entry = tk.Entry(width=6)
ur2entry.place(x = 280, y = 160)
dl1entry = tk.Entry(width=6)
dl1entry.place(x = 150, y = 290)
dl2entry = tk.Entry(width=6)
dl2entry.place(x = 180, y = 260)
dr1entry = tk.Entry(width=6)
dr1entry.place(x = 250, y = 290)
dr2entry = tk.Entry(width=6)
dr2entry.place(x = 280, y = 260)

# Importing images
weak_dominance_horizontal = tk.PhotoImage(file = 'weak_dominance_horizontal.gif')
strict_dominance_horizontal = tk.PhotoImage(file = 'strict_dominance_horizontal.gif')
weak_dominance_vertical = tk.PhotoImage(file = 'weak_dominance_vertical.gif')
strict_dominance_vertical = tk.PhotoImage(file = 'strict_dominance_vertical.gif')

# One global variable that will become of relevance for oddment calculations (which do not work in presence of strict dominance)
strict_dominance_exists = False


# ------------- Generating payoff matrixes for both players -------------

def insert_constant_sum_payoffs():
    if constantsumentry.get() == "" or constantsumentry.get() == None:
        sum = 0
    else:
        sum = float(constantsumentry.get())
    map_dict = {ul2entry:ul1entry, ur2entry:ur1entry, dl2entry:dl1entry, dr2entry:dr1entry}
    try:
        for column_entry_field in map_dict:
            new_value = sum - float(map_dict[column_entry_field].get())
            # The following block is to nicely display inserted values wihtout decimals if they are integers. For the calculations later on, all payoffs will be entered as floats into the payoff matrix
            if new_value.is_integer():
                new_value = int(new_value)  
            column_entry_field.insert(0, new_value)
    except ValueError:
        warning2 = tkinter.messagebox.showinfo(message="Please fill in all Row player fields.")
    

def get_payoffs():
    # Returns payoff matrix as a list in form: 
    # [ 
    #       [[Row-up-left, Row-up-right], [Row-down-left, Row-down-right]],
    #       [[Column-up-left, Column-up-right], [Column-down-left, Column-down-right]]
    # ]
    try:
        
        payoffs = [
        
            [
                [float(ul1entry.get()), float(ur1entry.get())],
                [float(dl1entry.get()), float(dr1entry.get())],
            ],
        
            [
                [float(ul2entry.get()), float(ur2entry.get())],
                [float(dl2entry.get()), float(dr2entry.get())],
            ]
        ]
        return payoffs
        
    except ValueError:
        return None
        
        
def flip(x): # Turns an index number of 0 or 1 into its opposite
    return ((x + 1) % 2)
        
        
def dominancechecker(checked1, alternative1, checked2, alternative2): # Checks a given strategy for strict or weak dominance over its alternative
    if (checked1 == alternative1) and (checked2 == alternative2):
        return 0
    if (checked1 > alternative1) and (checked2 > alternative2):
        global strict_dominance_exists 
        strict_dominance_exists = True
        return 2
    if (checked1 >= alternative1) and (checked2 >= alternative2):
        return 1
    return 0
    
    
def arrow_painter(list):
    
    if list[0] + list[1] > 0: # Tests if there is any dominance among U and D (the row player's strategies)
        if list[0] > 0:
            if list[0] == 1:
                i = weak_dominance_horizontal
            else: 
                i = strict_dominance_horizontal
            x1 = 20
            y1 = 140
        if list[1] > 0:
            if list[1] == 1:
                i = weak_dominance_horizontal
            else:
                i = strict_dominance_horizontal
            x1 = 20
            y1 = 275
        arrow1 = tk.Label(image=i)
        arrow1.place(x = x1, y = y1)
        
    if list[2] + list[3] > 0: # Tests if there is any dominance among L and R (the column player's strategies)
        if list[2] >0:
            if list[2] == 1:
                j = weak_dominance_vertical
            else:
                j = strict_dominance_vertical
            x2 = 150
            y2 = 5
        else:
            if list[3] == 1:
                j = weak_dominance_vertical
            else:
                j = strict_dominance_vertical
            x2 = 275
            y2 = 5
        arrow2 = tk.Label(image=j)
        arrow2.place(x = x2, y = y2)
    
    return
    
    
def nashequil(payoffs):
    number_of_pure_strategy_equilibria = 0
    row_player_payoff = None
    column_player_payoff = None
    for j in range(2):
        for k in range(2):
            if (payoffs[0][k][j] >= payoffs[0][flip(k)][j]) and (payoffs[1][k][j] >= payoffs[1][k][flip(j)]):
                    number_of_pure_strategy_equilibria += 1
                    if (j == 0) and (k == 0):
                        canvas.create_rectangle(135, 145, 220, 210, fill=None, width=5, outline="blue") # Up-Left
                        row_player_payoff = payoffs[0][0][0]
                        column_player_payoff = payoffs[1][0][0]
                    if (j== 0) and (k == 1):
                        canvas.create_rectangle(135, 245, 220, 310, fill=None, width=5, outline="blue") # Down-Left
                        row_player_payoff = payoffs[0][1][0]
                        column_player_payoff = payoffs[1][1][0]
                    if (j == 1) and (k == 0):
                        canvas.create_rectangle(235, 145, 320, 210, fill=None, width=5, outline="blue") # Up-Right
                        row_player_payoff = payoffs[0][0][1]
                        column_player_payoff = payoffs[1][0][1]
                    if (j == 1) and (k == 1):
                        canvas.create_rectangle(235, 245, 320, 310, fill=None, width=5, outline="blue") # Down-right
                        row_player_payoff = payoffs[0][1][1]
                        column_player_payoff = payoffs[1][1][1]
                
    return (number_of_pure_strategy_equilibria, row_player_payoff, column_player_payoff)
    
    
        
        
def calculate():
    payoffs = get_payoffs()
    if payoffs == None:
        msgwrng3 = "Please ensure all fields are filled before proceeding."
        warning3 = tkinter.messagebox.showinfo(message=msgwrng3)
        return
    
    dominance_results = [0, 0, 0, 0] # Preparing list of results of dominance checks
    # Running dominance checks for row player:
    for i in range(2):
        dominance_results[i] = dominancechecker(payoffs[0][i][0], payoffs[0][flip(i)][0], payoffs[0][i][1], payoffs[0][flip(i)][1])
    
    # Running dominance checks for column player
    for i in range(2):
        dominance_results[i + 2] = dominancechecker(payoffs[1][0][i], payoffs[1][0][flip(i)], payoffs[1][1][i], payoffs[1][1][flip(i)])
        
    # dominance_results is now a list of four variables where 2 reflects a strictly cominant strategy, 1 a weakly dominant one, and 0 a non-dominant strategy
    # These variables correspond to the following strategies: [Row-Up, Row-Down, Column-Left, Column-Right]
    if sum(dominance_results) > 0: # Calls the arrow_painter function (which displays dominance results graphically) if some dominance is present
        arrow_painter(dominance_results)
    results_of_nash_analysis = nashequil(payoffs) # Calls the function to check for Nash equilibria
        
    # The following part calculates the value of the game:   
    if results_of_nash_analysis[0] == 1: # Means there is only one pure strategy Nash equilibrium
        p = results_of_nash_analysis[1]
        q = results_of_nash_analysis[2]
        value_to_row_player = results_of_nash_analysis[1]
        value_to_column_player = results_of_nash_analysis[2]
    else:
        # This part calculates the probabilities with which the players should play
        # their pure strategies in a mixed strategy Nash equilibirum
        # Probability for row player playing U is p, for D 1-p
        # Probability for column player playing L is q, for R 1-q
        numerator_q = payoffs[0][1][1] - payoffs[0][0][1]
        denominator_q = payoffs[0][0][0] - payoffs[0][0][1] - payoffs[0][1][0] + payoffs[0][1][1]
        numerator_p = payoffs[1][1][1] - payoffs[1][1][0]
        denominator_p = payoffs[1][0][0] - payoffs[1][1][0] - payoffs[1][0][1] + payoffs[1][1][1]
        p = numerator_p / denominator_p
        q = numerator_q / denominator_q
        p_label = tk.Label(text = f"p = {round(p, 2)}")
        pm1_label = tk.Label(text = f"1-p = {round(1 - p, 2)}")
        q_label = tk.Label(text = f"q = {round(q, 2)}")
        qm1_label = tk.Label(text = f"1-q = {round(1 - q, 2)}")
        p_label.place(x = 350, y = 165)
        pm1_label.place(x = 350, y = 265)
        q_label.place(x = 155, y = 350)
        qm1_label.place(x = 255, y = 350)
        value_to_row_player = payoffs[0][0][0] * p + payoffs[0][1][0] * (1 - p)
        value_to_column_player = payoffs[1][0][0] * q + payoffs[1][0][1] * (1 - q)
        
    value_label = tk.Label(text = f"Values:\nRow player: {round(value_to_row_player, 2)}\nColumn player: {round(value_to_column_player, 2)}")
    value_label.place(x = 330, y = 380)

    return
    
    
calculatebutton = tk.Button(text="Calculate", command=calculate, width=10)
calculatebutton.place(x = 350, y = 10)        
        
zerosumbutton = tk.Button(text="Zero sum", command=insert_constant_sum_payoffs, width=10)
zerosumbutton.place(x = 20, y = 400)

constantsumentry = tk.Entry(width = 10)
constantsumentry.place(x = 110, y = 430)

constantsumbutton = tk.Button(text="Constant sum", command=insert_constant_sum_payoffs, width=10)
constantsumbutton.place(x = 110, y = 400)

# ---------------------------- Main program ----------------------------

window.mainloop()

File: 2x2gamesolver.c
#include <stdio.h>
#include <stdlib.h>

//Function prototypes
void asciiart(void);
void get_payoffs(float matrix[2][2][2]);
void print_overview(float matrix[2][2][2], int p);
int dominancechecker(float checked1, float alternative1, float checked2, float alternative2);
int flip (int x);


int main(void)
{
	asciiart();
	float payoffs[2][2][2]; //Generate empty payoff matrix
	get_payoffs(payoffs); //Call function to fill payoff matrix
	
	//Print overview table of payoff values
	printf("\n\nOverview over collected payoffs:\n\n");
    print_overview(payoffs, 0);
    printf("\nColumn player:\n------------\n");
    print_overview(payoffs, 1);
	
	
	int *dominance_results = malloc(sizeof(int) * 4); //allocate memory for four dominance variables:
	//A dominance variable is 2 if corresponding pure strategy is strictly dominant,
	//1 if weakly dominant,
	//0 otherwise
	
    //Running dominance checks for row player
    int i = 0;
    while (i < 2)
    {
        dominance_results[i] = dominancechecker(payoffs[0][i][0], payoffs[0][flip(i)][0], payoffs[0][i][1], payoffs[0][flip(i)][1]);
        i++;
    }
    
    //Running dominance checks for column player
    i = 0;
    while (i < 2)
    {
        dominance_results[i + 2] = dominancechecker(payoffs[1][0][i], payoffs[1][0][flip(i)], payoffs[1][1][i], payoffs[1][1][flip(i)]);
        i++;
    }
	
	char* pure_strategies[] = {"Up", "Down", "Left", "Right"}; //Array that assigns each numerical code for a pure strategy the name of that strategy
	
	//Print results of dominance check
	printf("\nResults of dominance check:\n\n");
    int* dominance_sum = malloc(sizeof(int)); //dominance_sum is used to keep track if any dominance exists (0 if there is none)
    if (dominance_sum == NULL)
        return 1;
    *dominance_sum = 0;
    i = 0;
	while (i < 4)
	{
		if (dominance_results[i] == 2)
        {
			printf("%s is strictly dominant\n", pure_strategies[i]);
            (*dominance_sum)++;
        }
		else if (dominance_results[i] == 1)
        {
			printf("%s is weakly dominant\n", pure_strategies[i]);
            (*dominance_sum)++;
        }
        i++;
	}
    if (*dominance_sum == 0)
        printf("No dominance detected\n");
	
    printf("\nPure strategy Nash equilibria:\n");
    
    int nash[2][2]; //Array that holds binary variables for pure strategies
    //Value to be 1 if that pure strategy is a Nash equilibrium, 0 otherwise
    int nashsum = 0; //sum of the Nash array
    for (int j1 = 0; j1 < 2; j1++)
    {
        for (int k1 = 0; k1 < 2; k1++)
        {
            if ((payoffs[0][k1][j1] >= payoffs[0][flip(k1)][j1]) && (payoffs[1][k1][j1] >= payoffs[1][k1][flip(j1)]))
                {
                    nash[k1][j1] = 1;
                    nashsum++;
                }
            else
                nash[k1][j1] = 0;
        }
    }
    
    if (nashsum == 0)
            printf("No pure strategy Nash equilibria detected.\n");
    else
    {
        for (int j2 = 0; j2 < 2; j2++)
        {
            for (int k2 = 0; k2 < 2; k2++)
            {
                if (nash[k2][j2] == 1)
                    printf("%s, %s\n", pure_strategies[k2], pure_strategies[j2 + 2]);
            }
        }
    }
    
    //Calculate mixed strategy Nash equilibria
    float numerator_p, numerator_q;
    float denominator_p, denominator_q;
    numerator_q = payoffs[0][1][1] - payoffs[0][0][1];
    denominator_q = payoffs[0][0][0] - payoffs[0][0][1] - payoffs[0][1][0] + payoffs[0][1][1];
    numerator_p = payoffs[1][1][1] - payoffs[1][1][0];
    denominator_p = payoffs[1][0][0] - payoffs[1][1][0] - payoffs[1][0][1] + payoffs[1][1][1];
    float p = numerator_p / denominator_p;
    float q = numerator_q / denominator_q;
    
    if ((0 < p) && (0 < q ) && ( p < 1) && (q < 1))
        printf("\nResults of mixed strategy analysis:\np: %f\nq: %f", p, q);
    else
        printf("\nNo mixed strategy Nash equlibria\n");
    
    //End of main function
    
    free(dominance_results);
    free(dominance_sum);
	return 0;
}


void asciiart(void)
{
	int number_of_lines = 11;
	char *lines[number_of_lines];
	lines[0] = "\n";
	lines[1] = "               L           R     \n";
	lines[2] = "\n";
	lines[3] = "   U            E           F        p\n";
	lines[4] = "               A           B          \n";
	lines[5] = "\n";
	lines[6] = "   D            G           H        (1-p)\n";
	lines[7] = "               C           D              \n";
	lines[8] = "\n";
	lines[9] = "               q          (1-q)      \n";
	lines[10] = "\n";

	for (int i = 0; i < number_of_lines; i++)
		printf("%s\n", lines[i]);
}

void get_payoffs(float matrix[2][2][2]) //This function fills the payoff matrix
{
	
	//Collect manual entry of payoffs for row player
	printf("A  B\n");
	printf("C  D\n");
	printf("Please enter payoffs for row player A, B, C, D, separated by commas: ");
	scanf("%f, %f, %f, %f", &matrix[0][0][0], &matrix[0][0][1], &matrix[0][1][0], &matrix[0][1][1]);
	
	//Ask if game is fixed sum (ask until answer is y, Y, n or N)
	char fixed_sum;
	while ( ((int) fixed_sum != 89) && ( (int) fixed_sum != 121) && ( (int) fixed_sum != 78) && ( (int) fixed_sum != 110) )
	{
	printf("\nFixed sum (y/n)? ");
	scanf(" %c", &fixed_sum);
	}
	
	if ( ( (int) fixed_sum == 89) || ( (int) fixed_sum == 121) )
	{	//If fixed-sum, automatically calculate column player payoffs as difference between sum and row player payoffs
		float sum;
		printf("What is that sum? ");
		scanf("%f", &sum);
		for (int i = 0; i < 2; i++)
        {
            for (int j = 0; j < 2; j++)
                matrix[1][i][j] = sum - matrix[0][i][j];
        }
	}	
	else if ( ( (int) fixed_sum == 78) || ( (int) fixed_sum == 110) )	
	{	//If not fixed sum, collect manual entry of column player payoffs
		printf("E  F\n");
		printf("G  H\n");
		printf("Please enter payoffs for column player E, F, G, H, separated by commas: ");
		scanf("%f, %f, %f, %f", &matrix[1][0][0], &matrix[1][0][1], &matrix[1][1][0], &matrix[1][1][1]);
	}
	
	
	return;
}

void print_overview(float matrix[2][2][2], int p)
{
    for (int x = 0; x < 2; x++)
    {
        for (int y = 0; y < 2; y++)
        {
            printf("%f       ", matrix[p][x][y]);
            if (y == 1)
                printf("\n");
        }
    }
}

int dominancechecker(float checked1, float alternative1, float checked2, float alternative2) //Checks a given strategy for strict or weak dominance over its alternative
{
	if ((checked1 > alternative1) && (checked2 > alternative2))
		return 2;
	else if ((checked1 >= alternative1) && (checked2 >= alternative2))
		return 1;
	else
		return 0;
}

int flip(int x) //Turns an index number of 0 or 1 into its opposite
{
    return ((x + 1) % 2);
}
